firmware/python_modules/sha2017/buttons.py

```diff
@@ -1,165 +1,175 @@

-

-def getCallback(button):

-	# Returns the currently attached callback function

-	global __num, __cb

-	if button < 0 or button >= __num:

-		raise ValueError("Invalid button!")

-	return __cb[-1][button]

-

-def pushMapping(newMapping=None):

-	global __cb

-	if newMapping == None:

-		newMapping = { BTN_UP: None, BTN_DOWN: None, BTN_LEFT: None, BTN_RIGHT: None, BTN_A: None, BTN_B: None, BTN_SELECT: None, BTN_START: None }

-		if machine.nvs_getint("system", 'factory_checked'):

-			newMapping[BTN_START] = __cbReboot

-	__cb.append(newMapping)

-

-def popMapping():

-	global __cb

-	if len(__cb) > 0:

-		__cb = __cb[:-1]

-	if len(__cb) < 1:

-		pushMapping()

-

-def rotate(value):

-	global __orientation

-	__orientation = value

-

-# ---

-__init()

+# SHA2017, Hackerhotel 2019 and Disobey 2020 button input wrapper
+# Versions of this file tailored for other badges should expose the same API
+
+import mpr121, _buttons, _mpr121mapping, system, machine
+
+# --- BUTTON CONSTANTS  ---
+BTN_A      = 0
+BTN_B      = 1
+BTN_START  = 2
+BTN_SELECT = 3
+BTN_DOWN   = 4
+BTN_RIGHT  = 5
+BTN_UP     = 6
+BTN_LEFT   = 7
+BTN_FLASH  = 8
+
+# Internal variables
+__num = 9
+__mprMap = _mpr121mapping.buttons
+__orientation = 0
+__cb = []
+
+# Internal functions
+def __cbReboot(pressed):
+    if pressed:
+        system.launcher()
+
+def __cb_btn_a(arg):
+    if __cb[-1][BTN_A]:
+        __cb[-1][BTN_A](arg)
+
+def __cb_btn_b(arg):
+    if __cb[-1][BTN_B]:
+        __cb[-1][BTN_B](arg)
+
+def __cb_btn_start(arg):
+    if __cb[-1][BTN_START]:
+        __cb[-1][BTN_START](arg)
+
+def __cb_btn_select(arg):
+    if __cb[-1][BTN_SELECT]:
+        __cb[-1][BTN_SELECT](arg)
+
+def __cb_btn_down(arg):
+    btn = BTN_DOWN
+    if __orientation == 90:
+        btn = BTN_LEFT
+    elif __orientation == 180:
+        btn = BTN_UP
+    elif __orientation == 270:
+        btn = BTN_RIGHT
+    if __cb[-1][btn]:
+        __cb[-1][btn](arg)
+
+def __cb_btn_right(arg):
+    btn = BTN_RIGHT
+    if __orientation == 90:
+        btn = BTN_DOWN
+    elif __orientation == 180:
+        btn = BTN_LEFT
+    elif __orientation == 270:
+        btn = BTN_UP
+    if __cb[-1][btn]:
+        __cb[-1][btn](arg)
+
+def __cb_btn_up(arg):
+    btn = BTN_UP
+    if __orientation == 90:
+        btn = BTN_RIGHT
+    elif __orientation == 180:
+        btn = BTN_DOWN
+    elif __orientation == 270:
+        btn = BTN_LEFT
+    if __cb[-1][btn]:
+        __cb[-1][btn](arg)
+
+
+def __cb_btn_left(arg):
+    btn = BTN_LEFT
+    if __orientation == 90:
+        btn = BTN_UP
+    elif __orientation == 180:
+        btn = BTN_RIGHT
+    elif __orientation == 270:
+        btn = BTN_DOWN
+    if __cb[-1][btn]:
+        __cb[-1][btn](arg)
+
+def __cb_btn_flash(arg):
+    if __cb[-1][BTN_FLASH]:
+        __cb[-1][BTN_FLASH](arg)
+
+def __init():
+    _buttons.register( 0,                    __cb_btn_flash  ) # The flash button is connected to the badge on GPIO 0

```

Comment:
Can we add the button to `__mprMap`?

Reply:
No, it's NOT part of the MPR121 input but a GPIO button instead. That is also why it is handled differently: different driver...